runtime/interp.c

```diff
@@ -873,52 +932,78 @@ value caml_interprete(code_t prog, asize_t prog_size)
+    raise_exception:
       Check_trap_barrier;
-      if (Caml_state->backtrace_active) {
+      if (domain_state->backtrace_active) {
         *--sp = (value)(pc - 1);
         caml_stash_backtrace(accu, sp, 0);
       }
     raise_notrace:
-      if ((char *) Caml_state->trapsp
-          >= (char *) Caml_state->stack_high - initial_sp_offset) {
-        Caml_state->external_raise = initial_external_raise;
-        Caml_state->extern_sp = (value *) ((char *) Caml_state->stack_high
-                                    - initial_sp_offset);
-        caml_callback_depth--;
-        return Make_exception_result(accu);
+      if (domain_state->trap_sp_off > 0) {
+        if (Stack_parent(domain_state->current_stack) == NULL) {
+          domain_state->external_raise = initial_external_raise;
+          domain_state->trap_sp_off = initial_trap_sp_off;
+          domain_state->current_stack->sp =
+            Stack_high(domain_state->current_stack) - initial_stack_words ;
+          return Make_exception_result(accu);
+        } else {
+          struct stack_info* old_stack = domain_state->current_stack;
+          struct stack_info* parent_stack = Stack_parent(old_stack);
+          value hexn = Stack_handle_exception(old_stack);
+          old_stack->sp = sp;
+          domain_state->current_stack = parent_stack;
+          sp = domain_state->current_stack->sp;
+          caml_free_stack(old_stack);
+
+          domain_state->trap_sp_off = Long_val(sp[0]);
+          extra_args = Long_val(sp[1]);
+          sp++;
+          sp[0] = accu;
+
+          accu = hexn;
+          pc = Code_val(accu);
+          env = accu;
+          goto check_stacks;
+        }
+      } else {
+        sp =
+           Stack_high(domain_state->current_stack) + domain_state->trap_sp_off;
+        pc = Trap_pc(sp);
+        domain_state->trap_sp_off = Long_val(Trap_link(sp));
+        env = sp[2];
+        extra_args = Long_val(sp[3]);
+        sp += 4;
       }
-      sp = Caml_state->trapsp;
-      pc = Trap_pc(sp);
-      Caml_state->trapsp = sp + Long_val(Trap_link_offset(sp));
-      env = sp[2];
-      extra_args = Long_val(sp[3]);
-      sp += 4;
       Next;
 
 /* Stack checks */
 
     check_stacks:
-      if (sp < Caml_state->stack_threshold) {
-        Caml_state->extern_sp = sp;
-        caml_realloc_stack(Stack_threshold / sizeof(value));
-        sp = Caml_state->extern_sp;
+      if (sp < Stack_threshold_ptr(domain_state->current_stack)) {
+        domain_state->current_stack->sp = sp;
+        if (!caml_try_realloc_stack(Stack_threshold / sizeof(value))) {
+          Setup_for_c_call; caml_raise_stack_overflow();
+        }
+        sp = domain_state->current_stack->sp;
       }
       /* Fall through CHECK_SIGNALS */
 
 /* Signal handling */
 
     Instruct(CHECK_SIGNALS):    /* accu not preserved */
-      if (caml_something_to_do) goto process_actions;
+      if (Caml_check_gc_interrupt(domain_state) ||
+          caml_check_for_pending_signals())
```

Comment:
In link with the polling behaviour changes, this can be made more efficient again. (Idem for `POPTRAP`.)